## VGG16

> only 3*3 kernel

> BatchNorm이 없던 시절에 만듦.

> Classifier 부분이 매우 무겁다.

In [1]:
import torchvision

c:\Users\gospe\miniconda3\Lib\site-packages\transformers\utils\generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(


In [2]:
model = torchvision.models.vgg16(pretrained = True)

c:\Users\gospe\miniconda3\Lib\site-packages\torchvision\models\_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
c:\Users\gospe\miniconda3\Lib\site-packages\torchvision\models\_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=VGG16_Weights.IMAGENET1K_V1`. You can also use `weights=VGG16_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


In [3]:
model

VGG(
  (features): Sequential(
    (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): ReLU(inplace=True)
    (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (3): ReLU(inplace=True)
    (4): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (5): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (6): ReLU(inplace=True)
    (7): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (8): ReLU(inplace=True)
    (9): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (10): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (11): ReLU(inplace=True)
    (12): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (13): ReLU(inplace=True)
    (14): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (15): ReLU(inplace=True)
    (16): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1

In [7]:
dir(model)

['T_destination',
 '__annotations__',
 '__call__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_apply',
 '_backward_hooks',
 '_backward_pre_hooks',
 '_buffers',
 '_call_impl',
 '_compiled_call_impl',
 '_forward_hooks',
 '_forward_hooks_always_called',
 '_forward_hooks_with_kwargs',
 '_forward_pre_hooks',
 '_forward_pre_hooks_with_kwargs',
 '_get_backward_hooks',
 '_get_backward_pre_hooks',
 '_get_name',
 '_is_full_backward_hook',
 '_load_from_state_dict',
 '_load_state_dict_post_hooks',
 '_load_state_dict_pre_hooks',
 '_maybe_warn_non_full_backward_hook',
 '_modules',
 '_named_members',
 '_non_persistent_buffers_se

In [4]:
model.classifier

Sequential(
  (0): Linear(in_features=25088, out_features=4096, bias=True)
  (1): ReLU(inplace=True)
  (2): Dropout(p=0.5, inplace=False)
  (3): Linear(in_features=4096, out_features=4096, bias=True)
  (4): ReLU(inplace=True)
  (5): Dropout(p=0.5, inplace=False)
  (6): Linear(in_features=4096, out_features=1000, bias=True)
)

In [9]:
model.avgpool

AdaptiveAvgPool2d(output_size=(7, 7))

In [1]:
print('hi')

hi


In [10]:
import torchvision.models as models
import torch.nn as nn

class VGG16(nn.Module):
    def __init__(self, class_num = 100): # 내 데이터셋의 클래스 수로 변경
        super(VGG16, self).__init__()
        self.class_num = class_num
        model = models.vgg16()
        model.classifier[-1] = nn.Linear(4096, class_num, bias = True) # VGG16의 마지막 레이어를 수정한다.
        self.model = model

    def forward(self, x):
        x = self.model(x)
        return x

In [11]:
model = VGG16()

In [12]:
model # output_features가 1,000에서 100으로 바뀜.

VGG16(
  (model): VGG(
    (features): Sequential(
      (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): ReLU(inplace=True)
      (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (3): ReLU(inplace=True)
      (4): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
      (5): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (6): ReLU(inplace=True)
      (7): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (8): ReLU(inplace=True)
      (9): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
      (10): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (11): ReLU(inplace=True)
      (12): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (13): ReLU(inplace=True)
      (14): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (15): ReLU(inplace=True)
      (16): Ma

위 classifier의 첫 FC_Layer의 입력 차원은 25,088(7 x 7 x 512)이다.

> AdaptAvgPool2D를 통해 (1 x 1 x feature_map)을 억지로 7 x 7로 늘림

> 어차피 1 x 1의 복사값이라 쓸 데 없는 연산임.

> so, Global Average Pooling을 통해 (1 x 1)을 그대로 사용

In [13]:
class VGG16(nn.Module):
    def __init__(self, class_num = 100): # 내 데이터셋의 클래스 수로 변경
        super(VGG16, self).__init__()
        self.class_num = class_num
        model = models.vgg16()
        model.avgpool = nn.AdaptiveAvgPool2d(output_size = (1,1))
        model.classifier = nn.Sequential(
            nn.Linear(512, 256, bias = True),
            nn.Relu(),
            nn.Dropout(),
            nn.Linear(256, 256, bias = True),
            nn.Relu(),
            nn.Dropout(),
            nn.Linear(256, class_num, bias = True)            
        )
        self.model = model

    def forward(self, x):
        x = self.model(x)
        return x

연산속도를 빠르게 하고 싶어서...

feature extraction의 일부분만을 활용하기

> + Maxpool 레이어 제거는 교안 참고

In [14]:
class VGG16(nn.Module):
    def __init__(self, class_num = 100): # 내 데이터셋의 클래스 수로 변경
        super(VGG16, self).__init__()
        self.class_num = class_num
        model = models.vgg16()
        model.features = model.features[:17] # 0~16번째 Layer만 사용 # 출력 차원을 잘 맞추어 classifier도 수정해야 한다!!
        model.avgpool = nn.AdaptiveAvgPool2d(output_size = (1,1))
        model.classifier = nn.Sequential(
            nn.Linear(256, 256, bias = True),
            nn.Relu(),
            nn.Dropout(),
            nn.Linear(256, 256, bias = True),
            nn.Relu(),
            nn.Dropout(),
            nn.Linear(256, class_num, bias = True)            
        )
        self.model = model

    def forward(self, x):
        x = self.model(x)
        return x